File: rs/src/display.rs
// Copyright © 2022 Mark Summerfield. All rights reserved.
// License: GPLv3

use crate::config::Config;
use crate::consts;
use crate::types::FileData;
use crate::util;
use num_format::{SystemLocale, ToFormattedString};
use std::{cmp::Ordering, collections::HashMap, time::Instant};

type NForLang = HashMap<String, usize>;

pub fn display_summary(
    file_data: Vec<FileData>,
    config: Config,
    t: Instant,
) {
    let lang_width = get_lang_width();
    let (total_for_lang, count_for_lang) = get_maps(&file_data);
    let totals = get_sorted_totals(&total_for_lang, config.sortbylines);
    for (lang, total) in totals {
        if let Some(count) = count_for_lang.get(lang) {
            display_summary_line(*count, total, lang, lang_width);
        }
    }
    let secs = t.elapsed().as_secs_f32();
    if secs > 0.1 {
        println!("{secs:.3} sec");
    }
}

fn get_lang_width() -> usize {
    consts::DATA_FOR_LANG
        .get()
        .values()
        .map(|v| v.name.len())
        .reduce(|acc, value| if acc >= value { acc } else { value })
        .unwrap_or(10)
}

fn get_maps(file_data: &[FileData]) -> (NForLang, NForLang) {
    let mut total_for_lang = NForLang::new();
    let mut count_for_lang = NForLang::new();
    for file_datum in file_data {
        total_for_lang
            .entry(file_datum.lang.clone())
            .and_modify(|lines| *lines += file_datum.lines)
            .or_insert(file_datum.lines);
        count_for_lang
            .entry(file_datum.lang.clone())
            .and_modify(|count| *count += 1)
            .or_insert(1);
    }
    (total_for_lang, count_for_lang)
}

fn get_sorted_totals(
    total_for_lang: &NForLang,
    sortbylines: bool,
) -> Vec<(&str, usize)> {
    let mut totals: Vec<(&str, usize)> = total_for_lang
        .iter()
        .map(|(lang, lines)| (lang.as_str(), *lines))
        .collect();
    if sortbylines {
        totals.sort_by(|a, b| {
            let (alang, alines) = a;
            let (blang, blines) = b;
            if alines != blines {
                alines.partial_cmp(blines).unwrap_or(Ordering::Equal)
            } else {
                alang
                    .to_lowercase()
                    .partial_cmp(&blang.to_lowercase())
                    .unwrap_or(Ordering::Equal)
            }
        });
    } else {
        totals.sort_by(|a, b| {
            let (alang, alines) = a;
            let (blang, blines) = b;
            alang
                .to_lowercase()
                .partial_cmp(&blang.to_lowercase())
                .unwrap_or_else(|| {
                    alines.partial_cmp(blines).unwrap_or(Ordering::Equal)
                })
        });
    }
    totals
}

fn display_summary_line(
    count: usize,
    total: usize,
    lang: &str,
    lang_width: usize,
) {
    let locale = SystemLocale::default().unwrap();
    let count_width = consts::FILE_COUNT_WIDTH;
    let lines_width = consts::LINE_COUNT_WIDTH;
    let s = if count == 1 { ' ' } else { 's' };
    if let Some(lang_data) = consts::DATA_FOR_LANG.get().get(&lang) {
        let count = count.to_formatted_string(&locale);
        let total = total.to_formatted_string(&locale);
        let name = &lang_data.name;
        println!(
            "{name:lang_width$} {count:>count_width$} file{s} \
            {total:>lines_width$} lines"
        );
    }
}

pub fn display_full(mut file_data: Vec<FileData>, config: Config) {
    let locale = SystemLocale::default().unwrap();
    let filename_width = get_filename_width(&file_data, config.maxwidth);
    let row_width = filename_width + 1 + consts::LINE_COUNT_WIDTH;
    let third = (filename_width / 3) - 1;
    let lines_width = consts::LINE_COUNT_WIDTH;
    let mut lang = String::new();
    let mut count = 0;
    let mut subtotal = 0;
    sort_file_data(&mut file_data, config.sortbylines);
    for file_datum in file_data {
        if lang.is_empty() || lang != file_datum.lang {
            (lang, count, subtotal) = display_new_lang(
                &lang,
                count,
                subtotal,
                row_width,
                &file_datum,
            );
        }
        let filename =
            util::elide(&file_datum.filename, third, filename_width);
        let lines = file_datum.lines.to_formatted_string(&locale);
        println!("{filename:filename_width$} {lines: >lines_width$}");
        subtotal += file_datum.lines;
        count += 1;
    }
    if !lang.is_empty() {
        display_subtotal(&lang, count, subtotal, row_width);
        #[cfg(windows)]
        println!("{}", "=".repeat(row_width));
        #[cfg(unix)]
        println!("{}", "━".repeat(row_width));
    }
}

fn get_filename_width(file_data: &[FileData], maxwidth: usize) -> usize {
    if let Some(width) = file_data
        .iter()
        .map(|f| f.filename.len())
        .reduce(|acc, value| if acc >= value { acc } else { value })
    {
        width
    } else {
        maxwidth
    }
    .min(maxwidth)
}

fn sort_file_data(file_data: &mut [FileData], sortbylines: bool) {
    if sortbylines {
        file_data.sort_by(|a, b| {
            if a.lang != b.lang {
                a.lang.partial_cmp(&b.lang).unwrap_or(Ordering::Equal)
            } else if a.lines != b.lines {
                a.lines.partial_cmp(&b.lines).unwrap_or(Ordering::Equal)
            } else {
                a.lang
                    .to_lowercase()
                    .partial_cmp(&b.lang.to_lowercase())
                    .unwrap_or(Ordering::Equal)
            }
        });
    } else {
        file_data.sort_by(|a, b| {
            if a.lang != b.lang {
                a.lang.partial_cmp(&b.lang).unwrap_or(Ordering::Equal)
            } else {
                a.filename
                    .to_lowercase()
                    .partial_cmp(&b.filename.to_lowercase())
                    .unwrap_or_else(|| {
                        a.lines
                            .partial_cmp(&b.lines)
                            .unwrap_or(Ordering::Equal)
                    })
            }
        });
    }
}

fn display_new_lang(
    lang: &str,
    count: usize,
    subtotal: usize,
    row_width: usize,
    file_datum: &FileData,
) -> (String, usize, usize) {
    let mut zeros = false;
    if !lang.is_empty() {
        display_subtotal(lang, count, subtotal, row_width);
        zeros = true;
    }
    let lang = &file_datum.lang;
    if let Some(lang_data) = consts::DATA_FOR_LANG.get().get(lang.as_str())
    {
        let name = format!(" {} ", lang_data.name);
        #[cfg(windows)]
        println!("{name:=^row_width$}");
        #[cfg(unix)]
        println!("{name:━^row_width$}");
    }
    if zeros {
        (lang.to_string(), 0, 0)
    } else {
        (lang.to_string(), count, subtotal)
    }
}

fn display_subtotal(
    lang: &str,
    count: usize,
    subtotal: usize,
    row_width: usize,
) {
    if let Some(lang_data) = consts::DATA_FOR_LANG.get().get(lang) {
        let name = lang_data.name;
        let locale = SystemLocale::default().unwrap();
        #[cfg(windows)]
        println!("{}", "-".repeat(row_width));
        #[cfg(unix)]
        println!("{}", "─".repeat(row_width));
        let s = if count == 1 { ' ' } else { 's' };
        let count = count.to_formatted_string(&locale);
        let subtotal = subtotal.to_formatted_string(&locale);
        let count_width = consts::FILE_COUNT_WIDTH;
        let lines_width = consts::LINE_COUNT_WIDTH;
        let numbers = format!(
            "{count:>count_width$} file{s} {subtotal:>lines_width$} lines"
        );
        let width = (row_width - numbers.len()).min(10000);
        println!("{name:<width$}{numbers}");
    }
}

File: go/display.go
// Copyright © 2022 Mark Summerfield. All rights reserved.
// License: Apache-2.0

package main

import (
	"fmt"
	"github.com/mark-summerfield/gong"
	"golang.org/x/text/message"
	"sort"
	"strings"
	"time"
	"unicode/utf8"
)

func displaySummary(dataForLang dataForLangMap, fileData []*fileDatum,
	sortByLines bool, d time.Duration) {
	totalForLang := make(map[string]int, len(dataForLang))
	countForLang := make(map[string]int, len(dataForLang))
	langWidth := 0
	for _, langData := range dataForLang {
		if size := utf8.RuneCountInString(langData.name); size > langWidth {
			langWidth = size
		}
	}
	lineFmt := fmt.Sprintf("%%-%ds %%%dd file%%s %%%dd lines\n", langWidth,
		fileCountWidth, lineCountWidth)
	for _, datum := range fileData {
		totalForLang[datum.lang] += datum.lines
		countForLang[datum.lang] += 1
	}
	langAndTotals := getLangAndTotals(totalForLang, sortByLines)
	out := message.NewPrinter(message.MatchLanguage("en"))
	for _, datum := range langAndTotals {
		count := countForLang[datum.lang]
		s := "s"
		if count == 1 {
			s = " "
		}
		langName := dataForLang[datum.lang].name
		out.Printf(lineFmt, langName, count, s, datum.total)
	}
	if d.Seconds() > 0.1 {
		fmt.Printf("%0.3f sec\n", d.Seconds())
	}
}

func getLangAndTotals(totalForLang map[string]int,
	sortByLines bool) []langAndTotal {
	langAndTotals := make([]langAndTotal, 0, len(totalForLang))
	for lang, total := range totalForLang {
		langAndTotals = append(langAndTotals, langAndTotal{lang, total})
	}
	sort.Slice(langAndTotals, func(i, j int) bool {
		if sortByLines {
			if langAndTotals[i].total == langAndTotals[j].total {
				return strings.ToLower(langAndTotals[i].lang) <
					strings.ToLower(langAndTotals[j].lang)
			}
			return langAndTotals[i].total < langAndTotals[j].total
		} else {
			return strings.ToLower(langAndTotals[i].lang) <
				strings.ToLower(langAndTotals[j].lang)
		}
	})
	return langAndTotals
}

func displayFull(dataForLang dataForLangMap, fileData []*fileDatum,
	sortByLines bool, maxWidth int) {
	filenameWidth := getWidth(fileData, maxWidth)
	rowWidth := filenameWidth + 1 + lineCountWidth
	lang := ""
	count := 0
	subtotal := 0
	filenameFmt := fmt.Sprintf("%%-%ds %%%dd\n", filenameWidth,
		lineCountWidth)
	out := message.NewPrinter(message.MatchLanguage("en"))
	sortFileData(fileData, sortByLines)
	for _, datum := range fileData {
		if lang == "" || lang != datum.lang {
			if lang != "" {
				displaySubtotal(dataForLang[lang].name, count, subtotal,
					rowWidth)
				count = 0
				subtotal = 0
			}
			name := " " + dataForLang[datum.lang].name + " "
			fmt.Println(gong.Centered(name, []rune(thick)[0], rowWidth))
		}
		lang = datum.lang
		filename := gong.ElideMiddle(datum.filename, filenameWidth)
		out.Printf(filenameFmt, filename, datum.lines)
		subtotal += datum.lines
		count++
	}
	if lang != "" {
		displaySubtotal(dataForLang[lang].name, count, subtotal,
			rowWidth)
		fmt.Println(strings.Repeat(thick, rowWidth))
	}
}

func sortFileData(fileData []*fileDatum, sortByLines bool) {
	sort.Slice(fileData, func(i, j int) bool {
		if sortByLines {
			if fileData[i].lang != fileData[j].lang {
				return fileData[i].lang < fileData[j].lang
			}
			if fileData[i].lines == fileData[j].lines {
				return strings.ToLower(fileData[i].filename) <
					strings.ToLower(fileData[j].filename)
			}
			return fileData[i].lines < fileData[j].lines
		} else {
			if fileData[i].lang != fileData[j].lang {
				return fileData[i].lang < fileData[j].lang
			}
			return strings.ToLower(fileData[i].filename) <
				strings.ToLower(fileData[j].filename)
		}
	})
}

func getWidth(fileData []*fileDatum, maxWidth int) int {
	width := 0
	for _, datum := range fileData {
		size := utf8.RuneCountInString(datum.filename)
		if size > width {
			width = size
			if maxWidth > 0 && width > maxWidth {
				return maxWidth
			}
		}
	}
	return width
}

func displaySubtotal(name string, count, subtotal, rowWidth int) {
	fmt.Println(strings.Repeat(thin, rowWidth))
	s := "s"
	if count == 1 {
		s = " "
	}
	numFmt := fmt.Sprintf("%%%dd file%s %%%dd lines", fileCountWidth, s,
		lineCountWidth)
	out := message.NewPrinter(message.MatchLanguage("en"))
	numbers := out.Sprintf(numFmt, count, subtotal)
	rowWidth -= utf8.RuneCountInString(numbers)
	rowFmt := fmt.Sprintf("%%-%ds%%s\n", rowWidth)
	fmt.Printf(rowFmt, name, numbers)
}
